### ISSUE TITLE
6.3 and later kernels fail to install

### ISSUE THREAD

**Issue by hax0rbana-adam** [linux-surface/linux-surface#1209]
The xrtx module fails to compile in Linux 6.3 and later.  I'm not sure where the xrtx module comes from, but I did find this [bug report](https://github.com/myriadrf/xtrx_linux_pcie_drv/issues/16) and a [pull request](https://github.com/myriadrf/xtrx_linux_pcie_drv/pull/17) that will fix it.

However, there are pull requests from over a year ago that haven't been merged into that repo, so I think it may have been abandoned.

I can't actually figure out where this driver came from. I don't see if here in this repo, but clearly it's on my freshly installed system. The pcie bus seems like a pretty important thing to have a driver for, but if it is something completely separate from the Linux-surface project, please let me know and I'll try to track it down again.

### Workaround
Downgrading to Linux to 6.2.10-surface works around this issue, but it would be nice to be able to switch back to using the `linux-image-surface` package so I get automatic updates again.

### Environment

- Hardware model: Surface Pro 6
- Kernel version: linux-image-6.4.7-surface
- Distribution: Debian 12

<details><summary>Error when installing latest kernel</summary>

```
# apt install -y linux-headers-surface linux-image-surface
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
linux-headers-surface is already the newest version (6.4.7-surface-3).
The following NEW packages will be installed:
  linux-image-6.4.7-surface linux-image-surface
0 upgraded, 2 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B/121 MB of archives.
After this operation, 139 MB of additional disk space will be used.
Selecting previously unselected package linux-image-6.4.7-surface.
(Reading database ... 349629 files and directories currently installed.)
Preparing to unpack .../linux-image-6.4.7-surface_6.4.7-surface-3_amd64.deb ...
Unpacking linux-image-6.4.7-surface (6.4.7-surface-3) ...
Selecting previously unselected package linux-image-surface.
Preparing to unpack .../linux-image-surface_6.4.7-surface-3_amd64.deb ...
Unpacking linux-image-surface (6.4.7-surface-3) ...
Setting up linux-image-6.4.7-surface (6.4.7-surface-3) ...
dkms: running auto installation service for kernel 6.4.7-surface.
Sign command: /lib/modules/6.4.7-surface/build/scripts/sign-file
Signing key: /var/lib/dkms/mok.key
Public certificate (MOK): /var/lib/dkms/mok.pub

Building module:
Cleaning build area...
make -j8 KERNELRELEASE=6.4.7-surface -C /lib/modules/6.4.7-surface/build M=/var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.2/build...(bad exit status: 2)
Error! Bad return status for module build on kernel: 6.4.7-surface (x86_64)
Consult /var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.2/build/make.log for more information.
Error! One or more modules failed to install during autoinstall.
Refer to previous errors for more information.
dkms: autoinstall for kernel: 6.4.7-surface failed!
run-parts: /etc/kernel/postinst.d/dkms exited with return code 11
update-initramfs: Generating /boot/initrd.img-6.4.7-surface
W: Possible missing firmware /lib/firmware/i915/dg2_huc_gsc.bin for module i915
Generating grub configuration file ...
Found background image: .background_cache.png
Found linux image: /boot/vmlinuz-6.4.7-surface
Found initrd image: /boot/initrd.img-6.4.7-surface
Found linux image: /boot/vmlinuz-6.2.10-surface
Found initrd image: /boot/initrd.img-6.2.10-surface
Found linux image: /boot/vmlinuz-6.1.0-10-amd64
Found initrd image: /boot/initrd.img-6.1.0-10-amd64
Warning: os-prober will not be executed to detect other bootable partitions.
Systems on them will not be added to the GRUB boot configuration.
Check GRUB_DISABLE_OS_PROBER documentation entry.
Adding boot menu entry for UEFI Firmware Settings ...
done
dpkg: error processing package linux-image-6.4.7-surface (--configure):
 installed linux-image-6.4.7-surface package post-installation script subprocess returned error exit status 1
dpkg: dependency problems prevent configuration of linux-image-surface:
 linux-image-surface depends on linux-image-6.4.7-surface (= 6.4.7-surface-3); however:
  Package linux-image-6.4.7-surface is not configured yet.

dpkg: error processing package linux-image-surface (--configure):
 dependency problems - leaving unconfigured
Errors were encountered while processing:
 linux-image-6.4.7-surface
 linux-image-surface
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

</details>

<details><summary>Important sections from make.log</summary>

```
DKMS make.log for xtrx-0.0.1+git20190320.5ae3a3e-3.2 for kernel 6.4.7-surface (x86_64)
Thu Aug  3 12:01:48 AM CDT 2023
make: Entering directory '/usr/src/linux-headers-6.4.7-surface'
warning: the compiler differs from the one used to build the kernel
  The kernel was built by: gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
  You are using:           gcc (Debian 12.2.0-14) 12.2.0
  CC [M]  /var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.2/build/xtrx.o
/var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.2/build/xtrx.c: In function ‘xtrxfd_mmap’:
/var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.2/build/xtrx.c:1075:31: error: assignment of read-only member ‘vm_flags’
 1075 |                 vma->vm_flags |= VM_LOCKED;
      |                               ^~
/var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.2/build/xtrx.c:1090:31: error: assignment of read-only member ‘vm_flags’
 1090 |                 vma->vm_flags |= VM_IO;
      |                               ^~
/var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.2/build/xtrx.c:1115:31: error: assignment of read-only member ‘vm_flags’
 1115 |                 vma->vm_flags |= VM_LOCKED;
      |                               ^~
In file included from ./include/linux/linkage.h:7,
                 from ./arch/x86/include/asm/cache.h:5,
                 from ./include/linux/cache.h:6,
                 from ./include/linux/time.h:5,
                 from ./include/linux/stat.h:19,
                 from ./include/linux/module.h:13,
                 from /var/lib/dkms/xtrx/0.0.1+git20190320.5ae3a3e-3.2/build/xtrx.c:22:
```

</details>

**Comment by qzed** [comment]
That's not something that we can fix in the kernel. You'll either have to install the module from the PR, wait for it to be fixed, or downgrade your kernel version. If you're unsure whether you need the module at all, you could also try to uninstall it.

**Comment by hax0rbana-adam** [comment]
Thank you for the reply, this helped me get to the bottom of it!

I thought this was something that was coming from the `linux-surface` apt repo, but I now realize that it comes from `xtrx-dkms` which is in the [stock Debian repos](https://packages.debian.org/bookworm/xtrx-dkms).  It turns out I don't actually need this kernel module, so I've uninstalled it and now I'm good to go. I'm closing this ticket and leaving some notes for the next person who runs into this or something similar.


For anyone else who might run into this issue, or really any issue where something is preventing the `linux-image-surface` package from being configured properly, here's how I figured out what was going on and determined that this is optional.

1. Find the package that caused the offending driver to get in there.  I did this with `apt list --installed *xtrx*`
2. Determine why it was installed (see below)

In my case, I found `xtrx-dkms` module was for a software defined radio (SDR) with `apt show xtrx-dkms`.  I also noticed that when I ran the `apt list ...` command earlier, it said the package was "[installed, automatic]", meaning I didn't install it, some other package pulled it in.

To find all the things that could have possibly pulled it in, I ran `apt-cache rdepends --installed xtrx-dkms`.  The only package the was listed was `libxtrxll0`.  Running `apt show libxtrxll0` confirms this, and I see it's only "recommended", not required.

Since `libxtrxll0` was also automatically installed, I decided to repeat this process until I figured out what I personally installed that caused this big chain reaction. In my case it turned out to be `gqrx-sdr`, in case you were curious.

I hope this info helps someone in the future. If so, give this comment a thumbs up. :slightly_smiling_face:
